File: apps/mobile-android/app/src/test/kotlin/br/com/tatame/feature/store/StoreFormatTest.kt
package br.com.tatame.feature.store

import br.com.tatame.core.network.dto.OrderStatuses
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * STO.12/13 — pure PT-BR store labels and the purchase state machine
 * (spec 009 fixed copy + stories 23–25).
 */
class StoreFormatTest {

    // ---- status chips (spec 009 fixed PT-BR copy) ------------------------

    @Test
    fun `status labels match the spec copy`() {
        assertEquals("Aguardando pagamento", StoreFormat.statusLabel(OrderStatuses.PENDING))
        assertEquals("Recebido", StoreFormat.statusLabel(OrderStatuses.PAID))
        assertEquals("Em andamento", StoreFormat.statusLabel(OrderStatuses.READY))
        assertEquals("Entregue", StoreFormat.statusLabel(OrderStatuses.DELIVERED))
        assertEquals("Cancelado", StoreFormat.statusLabel(OrderStatuses.CANCELED))
    }

    @Test
    fun `unknown status renders raw — registry growth never crashes`() {
        assertEquals("archived", StoreFormat.statusLabel("archived"))
    }

    // ---- money + lines ---------------------------------------------------

    @Test
    fun `price drops trailing cents like the prototype chips`() {
        assertEquals("R$ 389", StoreFormat.priceBRL(38_900))
        assertEquals("R$ 120,50", StoreFormat.priceBRL(12_050))
    }

    @Test
    fun `cta total is unit price times quantity`() {
        assertEquals(77_800L, StoreFormat.totalCents(38_900, 2))
    }

    @Test
    fun `pedido title and item line compose the card copy`() {
        assertEquals("#2431 · Kimono oficial", StoreFormat.pedidoTitle(2431, "Kimono oficial"))
        assertEquals("#2431", StoreFormat.pedidoTitle(2431, null))
        assertEquals("Tam M · 1 un · R$ 389", StoreFormat.itemLine("M", 1, 38_900))
        assertEquals("2 un · R$ 158", StoreFormat.itemLine(null, 2, 15_800)) // sizeless
    }

    @Test
    fun `pix sheet is addressed Pedido number produto`() {
        assertEquals(
            "Pedido #2431 · Kimono oficial",
            StoreFormat.pixSubtitle(2431, "Kimono oficial"),
        )
    }

    @Test
    fun `estoque line carries the retirada copy`() {
        assertEquals(
            "12 em estoque · retirada na recepção da academia",
            StoreFormat.estoqueLine(12),
        )
    }

    // ---- gallery derivation (spec — derivation, not schema) --------------

    @Test
    fun `gallery is the preset plus 2 catalog neighbors cycling`() {
        assertEquals(
            listOf("store-blue-purple", "store-teal-green", "store-orange-red"),
            StoreFormat.galleryPresets("store-blue-purple"),
        )
        assertEquals(
            listOf("store-pink-purple", "store-blue-purple", "store-teal-green"),
            StoreFormat.galleryPresets("store-pink-purple"),
        )
    }

    @Test
    fun `unknown preset starts the cycle at the catalog head`() {
        assertEquals(
            listOf("store-blue-purple", "store-teal-green", "store-orange-red"),
            StoreFormat.galleryPresets("store-future-preset"),
        )
    }

    // ---- purchase state machine (stories 23–25) --------------------------

    @Test
    fun `size is required exactly when the product defines sizes`() {
        val sizes = listOf("P", "M")
        assertFalse(StoreFormat.canBuy(sizes, selectedSize = null, stockQty = 5))
        assertTrue(StoreFormat.canBuy(sizes, selectedSize = "M", stockQty = 5))
        assertTrue(StoreFormat.canBuy(emptyList(), selectedSize = null, stockQty = 5))
    }

    @Test
    fun `esgotado disables the purchase regardless of size`() {
        assertFalse(StoreFormat.canBuy(emptyList(), selectedSize = null, stockQty = 0))
        assertFalse(StoreFormat.canBuy(listOf("M"), selectedSize = "M", stockQty = 0))
    }

    @Test
    fun `quantity clamps between 1 and the available stock`() {
        assertEquals(1, StoreFormat.clampQuantity(0, 12))
        assertEquals(12, StoreFormat.clampQuantity(99, 12))
        assertEquals(3, StoreFormat.clampQuantity(3, 12))
        // Esgotado (or oversold negative stock) floors at 1 — the CTA is
        // disabled anyway, the stepper never shows 0.
        assertEquals(1, StoreFormat.clampQuantity(5, 0))
    }
}

File: apps/mobile-rn/tests/features/store-format.test.ts
/**
 * Store pure logic (STO.10-11, spec 009 testing decisions): the detail
 * purchase state machine (sizeless/sized × stock levels → pill/stepper/CTA
 * state), the PT-BR status-chip label mapping, the fixed-copy formatters,
 * the deterministic gallery derivation and the histórico title split
 * (mensalidade vs avulso — order/event-origin charges have no competência).
 */

import { storeGalleryPresets } from '@tatame/design-system/native';
import { historyTitle } from '../../src/features/billing/format';
import {
  buyLabel,
  canCancelOrder,
  canResumePayment,
  galleryLabel,
  orderDateLine,
  orderItemLine,
  orderStatusChip,
  orderTitle,
  purchaseState,
  showsPickupNote,
  stockLine,
} from '../../src/features/store/format';
import { makeOrder, makeOrderItem, makeProductDetail } from '../helpers/store';

describe('purchase state machine (spec 009 stories 23-25)', () => {
  const sized = makeProductDetail(); // P/M/G/GG, 12 em estoque, R$ 389

  it('requires a size pill when the product defines sizes', () => {
    const without = purchaseState(sized, null, 1);
    expect(without.needsSize).toBe(true);
    expect(without.canBuy).toBe(false);

    const withSize = purchaseState(sized, 'M', 1);
    expect(withSize.canBuy).toBe(true);
  });

  it('buys immediately on sizeless products (tamanho is a choice, not a guess)', () => {
    const sizeless = makeProductDetail({ sizes: [], stockQty: 5 });
    const state = purchaseState(sizeless, null, 1);
    expect(state.needsSize).toBe(false);
    expect(state.canBuy).toBe(true);
  });

  it('caps the stepper at the available stock and floors it at 1', () => {
    const twoLeft = makeProductDetail({ sizes: [], stockQty: 2 });
    expect(purchaseState(twoLeft, null, 1).canDecrement).toBe(false);
    expect(purchaseState(twoLeft, null, 1).canIncrement).toBe(true);
    expect(purchaseState(twoLeft, null, 2).canIncrement).toBe(false);
    expect(purchaseState(twoLeft, null, 2).canBuy).toBe(true);
    // Beyond-stock quantity is never purchasable (story 24).
    expect(purchaseState(twoLeft, null, 3).canBuy).toBe(false);
  });

  it('flags esgotado at zero stock and disables everything', () => {
    const soldOut = purchaseState(makeProductDetail({ stockQty: 0 }), 'M', 1);
    expect(soldOut.soldOut).toBe(true);
    expect(soldOut.canBuy).toBe(false);
    expect(soldOut.canIncrement).toBe(false);
  });

  it('multiplies the CTA total by the quantity (price snapshot × qty)', () => {
    expect(purchaseState(sized, 'M', 1).ctaLabel).toBe(
      'Comprar com Pix · R$ 389,00',
    );
    expect(purchaseState(sized, 'M', 3).ctaLabel).toBe(
      'Comprar com Pix · R$ 1.167,00',
    );
    expect(buyLabel(38_900, 2)).toBe('Comprar com Pix · R$ 778,00');
  });
});

describe('order status chips (spec 009 PT-BR label registry)', () => {
  it.each([
    ['pending', 'Aguardando pagamento', 'warning'],
    ['paid', 'Recebido', 'success'],
    ['ready', 'Em andamento', 'brand'],
    ['delivered', 'Entregue', 'neutral'],
    ['canceled', 'Cancelado', 'danger'],
  ] as const)('%s → "%s" (%s)', (status, label, tone) => {
    expect(orderStatusChip(status)).toEqual({ label, tone });
  });

  it('shows the retirada note only while there is something to pick up', () => {
    expect(showsPickupNote('paid')).toBe(true);
    expect(showsPickupNote('ready')).toBe(true);
    expect(showsPickupNote('pending')).toBe(false);
    expect(showsPickupNote('delivered')).toBe(false);
    expect(showsPickupNote('canceled')).toBe(false);
  });

  it('cancels and resumes payment on pending orders only (story 28)', () => {
    expect(canCancelOrder(makeOrder())).toBe(true);
    expect(canResumePayment(makeOrder())).toBe(true);
    expect(canCancelOrder(makeOrder({ status: 'paid' }))).toBe(false);
    expect(
      canResumePayment(makeOrder({ status: 'paid', chargeId: null })),
    ).toBe(false);
    expect(canResumePayment(makeOrder({ chargeId: null }))).toBe(false);
  });
});

describe('fixed-copy formatters', () => {
  it('renders the order lines per the prototype', () => {
    expect(orderTitle(makeOrder())).toBe('Pedido #2431');
    expect(orderItemLine(makeOrderItem())).toBe(
      'Kimono oficial Horizonte · Tam M · 1 un',
    );
    expect(orderItemLine(makeOrderItem({ size: null, quantity: 2 }))).toBe(
      'Kimono oficial Horizonte · 2 un',
    );
    expect(orderDateLine('2026-08-08T14:00:00.000Z')).toBe('Feito em 08/08');
    expect(stockLine(12)).toBe(
      '12 em estoque · retirada na recepção da academia',
    );
    expect(galleryLabel(1)).toBe('Foto 2 de 3');
  });
});

describe('gallery derivation (spec 009 — derivation, not schema)', () => {
  it('yields the own preset plus the 2 catalog neighbors, wrap-around', () => {
    expect(storeGalleryPresets('store-blue-purple')).toEqual([
      'store-blue-purple',
      'store-teal-green',
      'store-orange-red',
    ]);
    expect(storeGalleryPresets('store-pink-purple')).toEqual([
      'store-pink-purple',
      'store-blue-purple',
      'store-teal-green',
    ]);
  });

  it('falls back to the catalog head on unknown slugs (catalog may lag the server)', () => {
    expect(storeGalleryPresets('store-future-preset')[0]).toBe(
      'store-blue-purple',
    );
    expect(storeGalleryPresets(null)[0]).toBe('store-blue-purple');
  });
});

describe('carteira histórico title (spec 009 story 29)', () => {
  it('keeps the mensalidade label for plan charges (competência present)', () => {
    expect(
      historyTitle({ periodStart: '2026-07-01', paidAt: '2026-07-08' }),
    ).toBe('Mensalidade · julho');
  });

  it('labels order/event payments (no competência) as pagamento avulso', () => {
    expect(
      historyTitle({ periodStart: null, paidAt: '2026-08-08T14:00:00.000Z' }),
    ).toBe('Pagamento avulso');
  });
});
